### PR TITLE
Mats/POST for query api CERES-426

### PR DESCRIPTION
**Pull request by matsboman**
# Resolves

[CERES-426](https://jira.rax.io/browse/CERES-426)

# What

Create a POST /api/query which aligns with the OpenTsdb query api so grafana can pull data from the endpoint.

# How

Created a POST /api/query endpoint.
[Swagger schema](https://github.com/Rackspace-Segment-Support/ceres-docs-internal/blob/master/proposals/opentsdb-query-api-compatibility/opentsdb-subset.yaml)

## How to test

JUnit tests are implemented.

### CODE REVIEW

**Review by rohitsngh27 — CHANGES_REQUESTED** [review]
@matsboman I think you need to handle the case when `TsdbQueryRequest.downsample` is empty i.e., the case of querying raw data.

File: src/main/java/com/rackspace/ceres/app/services/MetadataService.java
Comment (harageth) [inline src/main/java/com/rackspace/ceres/app/services/MetadataService.java:234]: Why do we need to split on | ?

File: src/main/java/com/rackspace/ceres/app/services/MetadataService.java
Comment (harageth) [inline src/main/java/com/rackspace/ceres/app/services/MetadataService.java:241]: Same here with '-'?

File: src/main/java/com/rackspace/ceres/app/services/QueryService.java
Comment (harageth) [inline src/main/java/com/rackspace/ceres/app/services/QueryService.java:135]: This falls in line with how we are handling flux in our other API's so its fine. But it feels wrong. 

The API service should have a clear thing that it is doing, which is delegating responsibility to some other @Service and only handling API related things. Most of this data manipulation feels like it should be happening somewhere else, and stringing along all of these lambdas like this make it difficult to get whats going on at a glance. Mostly this comment is to start a larger discussion about this and maybe reconsider this style that we are using (maybe this is just the flux way).

File: src/main/java/com/rackspace/ceres/app/services/MetadataService.java
Comment (matsboman) [reply src/main/java/com/rackspace/ceres/app/services/MetadataService.java:234]: We only support "type", "literal_or" but you have a point maybe I need to enforce it?

File: src/main/java/com/rackspace/ceres/app/services/MetadataService.java
Comment (matsboman) [reply src/main/java/com/rackspace/ceres/app/services/MetadataService.java:241]: Need to parse out the granularity width and aggregator

File: src/main/java/com/rackspace/ceres/app/services/MetadataService.java
Comment (rohitsngh27) [reply src/main/java/com/rackspace/ceres/app/services/MetadataService.java:234]: I think splitting would be required in case of literal_or filter as there can be `|` separated values that we need to or.

File: src/main/java/com/rackspace/ceres/app/services/QueryService.java
Comment (rohitsngh27) [inline src/main/java/com/rackspace/ceres/app/services/QueryService.java:133]: Can you rename this to something like `queryTsdb` as this method is doing processing for both `raw` and `downsampled` data.

File: src/main/java/com/rackspace/ceres/app/model/TsdbQueryRequest.java
Comment (rohitsngh27) [inline src/main/java/com/rackspace/ceres/app/model/TsdbQueryRequest.java:12]: I think instead of creating this as Map of String we can create new class as `Filters` with the fields `type` which can again be an Enum having `literal_or` right now, `tagk` and `filter` as defined in http://opentsdb.net/docs/build/html/api_http/query/index.html

File: src/main/java/com/rackspace/ceres/app/services/MetadataService.java
Comment (rohitsngh27) [inline src/main/java/com/rackspace/ceres/app/services/MetadataService.java:227]: With reference to my comment in https://github.com/racker/ceres/pull/34/files#r607837075 I think the logic will be change here too.

File: src/main/java/com/rackspace/ceres/app/model/TsdbQueryRequest.java
Comment (rohitsngh27) [reply src/main/java/com/rackspace/ceres/app/model/TsdbQueryRequest.java:12]: @harageth What's your say on this one?

File: src/main/java/com/rackspace/ceres/app/services/QueryService.java
Comment (matsboman) [reply src/main/java/com/rackspace/ceres/app/services/QueryService.java]: done

File: src/main/java/com/rackspace/ceres/app/model/TsdbQueryRequest.java
Comment (matsboman) [reply src/main/java/com/rackspace/ceres/app/model/TsdbQueryRequest.java]: done

File: src/main/java/com/rackspace/ceres/app/services/MetadataService.java
Comment (matsboman) [reply src/main/java/com/rackspace/ceres/app/services/MetadataService.java:228]: done

**Review by rohitsngh27 — APPROVED** [review]
Changes seems fine to me but let's wait for review from @harageth before merging.

**Review by harageth — APPROVED** [review]

**Pull request merged**
